src/AggregateFunctions/AggregateFunctionGroupSortedArray.h

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include <Columns/ColumnArray.h>
+#include <DataTypes/DataTypeArray.h>
+
+#include <AggregateFunctions/IAggregateFunction.h>
+#include "AggregateFunctionGroupSortedArrayData.h"
+
+namespace DB
+{
+template <typename TT, bool is_plain_column>
+inline TT readItem(const IColumn * column, Arena * arena, size_t row)
+{
+    if constexpr (std::is_same_v<TT, StringRef>)
+    {
+        if constexpr (is_plain_column)
+        {
+            StringRef str = column->getDataAt(row);
+            auto ptr = arena->alloc(str.size);
+            std::copy(str.data, str.data + str.size, ptr);
+            return StringRef(ptr, str.size);
+        }
+        else
+        {
+            const char * begin = nullptr;
+            return column->serializeValueIntoArena(row, *arena, begin);
+        }
+    }
+    else
+    {
+        return column->getUInt(row);
+    }
+}
+
+template <bool is_plain_column, typename T, bool is_weighted>
+class AggregateFunctionGroupSortedArray : public IAggregateFunctionDataHelper<
+                                              AggregateFunctionGroupSortedArrayData<T, is_weighted>,
+                                              AggregateFunctionGroupSortedArray<is_plain_column, T, is_weighted>>
+{
+protected:
+    using State = AggregateFunctionGroupSortedArrayData<T, is_weighted>;
+
+    UInt64 threshold;
+    DataTypePtr & input_data_type;
+    mutable std::mutex mutex;
+
+    static void deserializeAndInsert(StringRef str, IColumn & data_to);
+
+public:
+    AggregateFunctionGroupSortedArray(UInt64 threshold_, UInt64 /*load_factor*/, const DataTypes & argument_types_, const Array & params)
+        : IAggregateFunctionDataHelper<AggregateFunctionGroupSortedArrayData<T, is_weighted>, AggregateFunctionGroupSortedArray>(
+            argument_types_, params)
+        , threshold(threshold_)
+        , input_data_type(this->argument_types[0])
+    {
+    }
+
+    void create(AggregateDataPtr place) const override
+    {
+        IAggregateFunctionDataHelper<AggregateFunctionGroupSortedArrayData<T, is_weighted>, AggregateFunctionGroupSortedArray>::create(
+            place);
+        this->data(place).threshold = threshold;
+    }
+
+    String getName() const override { return is_weighted ? "groupSortedArrayWeighted" : "groupSortedArray"; }
+
+    DataTypePtr getReturnType() const override { return std::make_shared<DataTypeArray>(input_data_type); }
+
+    bool allocatesMemoryInArena() const override
+    {
+        if constexpr (std::is_same_v<T, StringRef>)
+            return true;
+        else
+            return false;
+    }
+
+    void add(AggregateDataPtr __restrict, const IColumn **, size_t, Arena *) const override { }
+
+    void addBatchSinglePlace(
```

Comment:
Using addBatchSinglePlace() requires to calculate different agregations and combine all of them in the final result. I am also adding other optimizations like reading reading the first values (if weighted) only for the final results of each aggregation. It saves a lot of CPU when reading strings.

Reply:
I'm confused. What's the real benefit of implementing `addBatchSinglePlace`?
Having `addBatchSinglePlace` opens up the posibility of working with a full batch in interesting ways, like finding the batch top N elements and keeping only those (with a partial sort) and things like that, but if you are iterating over all elements and calling `add` then you could use the default implementation and avoid complexity, right?